### PR TITLE
Add BATS_SUITE_TMPDIR / BATS_FILE_TMPDIR / BATS_TEST_TMPDIR

### PR DESCRIPTION
**Pull request by rockandska**
- [X] I have reviewed the [Contributor Guidelines][contributor].
- [X] I have reviewed the [Code of Conduct][coc] and agree to abide by it

[contributor]: https://github.com/bats-core/bats-core/blob/master/docs/CONTRIBUTING.md
[coc]:         https://github.com/bats-core/bats-core/blob/master/docs/CODE_OF_CONDUCT.md

BATS_SUITE_TMPDIR : tmp directory specific to the suite
BATS_FILE_TMPDIR : tmp directory specific to the test file
BATS_TEST_TMPDIR : tmp directory specific to the test

Allow the users to use those directories to store files necessary in tests depending on if they want to share them across suite, test file, tests or not.

Solve #283 

To summarize :
```
BATS_TMPDIR
  |_ BATS_RUN_TMPDIR
    |_ BATS_SUITE_TMPDIR
      |_ BATS_FILE_TMPDIR
        |_ BATS_TEST_TMPDIR
```

### CODE REVIEW

**Comment by martin-schulze-vireso** [comment]
Currently, there is only one suite per run. This makes the suite and run tmpdir somewhat redundant, so I'd say we should avoid one of them. Unfortunately, I'd find suite, file, test more uniform than run, file, test. However, the `BATS_RUN_TMPDIR` might already have users because it has been released, albeit undocumented.

**Comment by rockandska** [comment]
> Currently, there is only one suite per run.

Indeed.

> This makes the suite and run tmpdir somewhat redundant, so I'd say we should avoid one of them.

BATS_RUN_TMPDIR ans BATS_SUITE_TMPDIR have different purpose. BATS_RUN_TMPDIR is intend to be used internally by bats (store SRC, logs, etc..) when BATS_SUITE_TMPDIR is intend to be used by users in tests (will be used in a future `before_all` functionality) to store temp files common to all tests.

> Unfortunately, I'd find suite, file, test more uniform than run, file, test. However, the `BATS_RUN_TMPDIR` might already have users because it has been released, albeit undocumented.

The actual design in this PR will not break tests if users actually using it.

**Comment by martin-schulze-vireso** [comment]
> BATS_RUN_TMPDIR ans BATS_SUITE_TMPDIR have different purpose. BATS_RUN_TMPDIR is intend to be used internally by bats (store SRC, logs, etc..) when BATS_SUITE_TMPDIR is intend to be used by users in tests (will be used in a future `before_all` functionality) to store temp files common to all tests.

Ahh, that makes sense. I thought they all were for internal usage only. Are you aware of https://github.com/bats-core/bats-file#working-with-temporary-directories? There seems to be some overlap.

**Comment by rockandska** [comment]
> Are you aware of https://github.com/bats-core/bats-file#working-with-temporary-directories? There seems to be some overlap.

Was not aware of `temp_make` but not a big fan of this for those reasons :
- users are the ones who need to create them (more code to write)
- users are the ones who need to manage the deletion (more code to write)
- users need to use additional env var to preserve temporary files (this PR implementation use the command-line parameter already documented and available in `--help` of bats and keep a logic directory structure as describe in the PR description)
- users need to add additional library totally uncorrelated from the core

File: libexec/bats-core/bats-exec-file [hunk 1]
```diff
@@ -49,6 +49,12 @@ fi
 
 BATS_TEST_FILENAME="$filename"
 
+if ! BATS_FILE_TMPDIR=$(mktemp -d "${BATS_SUITE_TMPDIR}/bats-file-tmpdir-XXXXXX");then
```
Comment (martin-schulze-vireso) [inline libexec/bats-core/bats-exec-file:52]: I think it would be easier for debugging/navigation to have the filename as part of the tmpdir. Maybe `${BATS_SUITE_TMPDIR}/file-tmpdir-$(basename "${BATS_FILENAME}")-XXXX`.

File: libexec/bats-core/bats-exec-test [hunk 2]
```diff
@@ -43,6 +43,12 @@ elif [[ ! -f "$BATS_TEST_FILENAME" ]]; then
   exit 1
 fi
 
+if ! BATS_TEST_TMPDIR=$(mktemp -d "${BATS_FILE_TMPDIR}/bats-test-tmpdir-XXXXXX");then
```
Comment (martin-schulze-vireso) [inline libexec/bats-core/bats-exec-test:46]: Likewise to the comment on BATS_FILE_TMPDIR it is more readable to have `${BATS_FILE_TMPDIR}/test-tmpdir-${BATS_TESTNAME}-XXXX`.

File: libexec/bats-core/bats-exec-suite [hunk 3]
```diff
@@ -120,6 +120,12 @@ if [[ "${test_count}" == 0 ]]; then
   exit
 fi
 
+if ! BATS_SUITE_TMPDIR=$(mktemp -d "${BATS_RUN_TMPDIR}/bats-suite-tmpdir-XXXXXX");then
```
Comment (martin-schulze-vireso) [inline libexec/bats-core/bats-exec-suite:123]: I think the `$BATS_RUN_TMPDIR` prefix gives us enough scope to omit the bats- prefix on lower levels, so `${BATS_RUN_TMPDIR}/suite-tmpdir-XXXXX` should be enough.

File: test/junit-formatter.bats [hunk 4]
```diff
@@ -103,8 +102,7 @@ TESTSUITES_REGEX="<testsuites time=\"$FLOAT_REGEX\">"
 }
 
 @test "junit formatter as report formatter creates report.xml" {
-  make_bats_test_suite_tmpdir
-  cd "$BATS_TEST_SUITE_TMPDIR" # don't litter sources with output files
+  cd "$BATS_TEST_TMPDIR" # don't litter sources with output files
```
Comment (martin-schulze-vireso) [inline test/junit-formatter.bats:105]: I know it is not really part of this PR but could you also take care of the `tab in filename` test? It litters source with the transformed file and should be trivially fixable with the new tmpdirs.

**Review by martin-schulze-vireso — COMMENTED** [review]
Only some minor issues.

File: test/junit-formatter.bats (hunk 4 above)
Comment (rockandska) [reply test/junit-formatter.bats:105]: I don't get what you would like to see to be done.
If it is not related directly to this PR, why not address this in another one ?
Always better to split PR for different purpose

File: test/bats.bats [hunk 5]
```diff
@@ -656,8 +650,7 @@ END_OF_ERR_MSG
 }
```
Comment (martin-schulze-vireso) [inline test/bats.bats:650]:
```suggestion
  cp "$BATS_TEST_TMPDIR/tab in filename.bats" "$FIXTURE_ROOT/tab"$'\t'"in filename.bats"
  bats "$BATS_TEST_TMPDIR/tab"$'\t'"in filename.bats"
}
```

I meant to change this, which is very similar to your change below and profits from your newly added variable.

Comment (martin-schulze-vireso) [inline test/bats.bats:650]: I meant you to change this, which is very similar to your change below and profits from your newly added variable:

```
  cp "$BATS_TEST_TMPDIR/tab in filename.bats" "$FIXTURE_ROOT/tab"$'\t'"in filename.bats"
  bats "$BATS_TEST_TMPDIR/tab"$'\t'"in filename.bats"
}
```

File: test/junit-formatter.bats (hunk 4 above)
Comment (martin-schulze-vireso) [reply test/junit-formatter.bats:105]: See my new comment. I think it is very fitting for this PR.

File: test/bats.bats (hunk 5 above)
Comment (rockandska) [reply test/bats.bats:650]: Should be fixed now

**Comment by rockandska** [comment]
My last concern, and seems directly related to the tests themself, is some tmp directory still here at the of tests
```
ls -1 tmp/
bats-run-test-tmpdir-2081-bOOCIO
bats-run-test-tmpdir-2196-LlHPJF
bats-run-test-tmpdir-2254-mKmbdo
bats-run-test-tmpdir-9566-OcLJnk
bats-run-test-tmpdir-9633-NNJAin
bats-run-test-tmpdir-9715-MLHeea
tmp.gfkgam
tmp.mgfpCl
tmp.nDgkNa
```

**Review by martin-schulze-vireso — APPROVED** [review]

**Comment by martin-schulze-vireso** [comment]
Thanks for the cleanup. I'll need to sign these commits to merge them. If you can, please do so yourself to reduce barriers.

**Comment by rockandska** [comment]
> Thanks for the cleanup. I'll need to sign these commits to merge them. If you can, please do so yourself to reduce barriers.

done
